### ISSUE TITLE
'cygheap base mismatch detected' when installing on windows 10 enterprise

### ISSUE THREAD

**Issue by magynhard** [oneclick/rubyinstaller2#111]
This project is for Ruby version 2.4 and newer.
For Ruby versions < 2.4 please file an issue [here]( https://github.com/oneclick/rubyinstaller/issues/new ).

## What problems are you experiencing?
When installing the MSYS2 toolings, i get unexpected errors.

Platform: 
Ruby 2.4.4-1 x64 (with Devkit)
Win10 Enterprise

```
 _____       _           _____           _        _ _         ___
|  __ \     | |         |_   _|         | |      | | |       |__ \
| |__) |   _| |__  _   _  | |  _ __  ___| |_ __ _| | | ___ _ __ ) |
|  _  / | | | '_ \| | | | | | | '_ \/ __| __/ _` | | |/ _ \ '__/ /
| | \ \ |_| | |_) | |_| |_| |_| | | \__ \ || (_| | | |  __/ | / /_
|_|  \_\__,_|_.__/ \__, |_____|_| |_|___/\__\__,_|_|_|\___|_||____|
                    __/ |           _
                   |___/          _|_ _  __   | | o __  _| _     _
                                   | (_) |    |^| | | |(_|(_)\^/_>

   1 - MSYS2 base installation
   2 - MSYS2 system update (optional)
   3 - MSYS2 and MINGW development toolchain

Which components shall be installed? If unsure press ENTER [1,2,3]

> sh -lc true
      1 [main] sh (12936) C:\Ruby24-x64\msys64\usr\bin\sh.exe: *** fatal error - cygheap base mismatch detected - 0x2F2F410/0x397F410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
      0 [main] sh 13440 fork: child -1 - forked process 12936 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
/etc/profile: fork: retry: Resource temporarily unavailable
      1 [main] sh (12448) C:\Ruby24-x64\msys64\usr\bin\sh.exe: *** fatal error - cygheap base mismatch detected - 0x2F2F410/0x38FF410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
1057375 [main] sh 13440 fork: child -1 - forked process 12448 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
/etc/profile: fork: retry: Resource temporarily unavailable
3107125 [main] sh 13440 fork: child -1 - forked process 11648 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
/etc/profile: fork: retry: Resource temporarily unavailable
      2 [main] sh (12380) C:\Ruby24-x64\msys64\usr\bin\sh.exe: *** fatal error - cygheap base mismatch detected - 0x2F2F410/0x348F410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
7181690 [main] sh 13440 fork: child -1 - forked process 12380 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
/etc/profile: fork: retry: Resource temporarily unavailable
      3 [main] sh (11616) C:\Ruby24-x64\msys64\usr\bin\sh.exe: *** fatal error - cygheap base mismatch detected - 0x2F2F410/0x395F410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
15272536 [main] sh 13440 fork: child -1 - forked process 11616 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
/etc/profile: fork: Resource temporarily unavailable
Download http://repo.msys2.org/distrib/x86_64/msys2-x86_64-20161025.exe
```

When try to install ruby-debug-ide with `gem install ruby-debug-ide`, which needs to be built native, similar errors occur:
```
Fetching: ruby-debug-ide-0.6.1.gem (100%)
Temporarily enhancing PATH for MSYS/MINGW...
Building native extensions.  This could take a while...
      1 [main] make (3296) C:\Ruby24-x64\msys64\usr\bin\make.exe: *** fatal error - cygheap base mismatch detected - 0x3C5F410/0x37CF410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
      0 [main] make 12976 fork: child -1 - forked process 3296 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
      1 [main] make (7344) C:\Ruby24-x64\msys64\usr\bin\make.exe: *** fatal error - cygheap base mismatch detected - 0x32FF410/0x347F410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
      0 [main] make 4784 fork: child -1 - forked process 7344 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
      1 [main] make (12368) C:\Ruby24-x64\msys64\usr\bin\make.exe: *** fatal error - cygheap base mismatch detected - 0x41BF410/0x37BF410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
      0 [main] make 8532 fork: child -1 - forked process 12368 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
      1 [main] make (668) C:\Ruby24-x64\msys64\usr\bin\make.exe: *** fatal error - cygheap base mismatch detected - 0x376F410/0x41FF410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
      0 [main] make 11272 fork: child -1 - forked process 668 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
ERROR:  Error installing ruby-debug-ide:
        ERROR: Failed to build gem native extension.

    current directory: C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/ruby-debug-ide-0.6.1/ext
C:/Ruby24-x64/bin/ruby.exe mkrf_conf.rb
Installing base gem
Building native extensions.  This could take a while...
Building native extensions.  This could take a while...
ERROR: Failed to build gem native extension.

    current directory: C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/debase-0.2.2/ext
C:/Ruby24-x64/bin/ruby.exe -r ./siteconf20180524-13360-136jom.rb extconf.rb
checking for vm_core.h... no
checking for version.h... no
checking for vm_core.h... no
checking for version.h... no
**************************************************************************
No source for ruby-2.4.4-p296 (revision 63013) provided with
debase-ruby_core_source gem. Falling back to ruby-2.4.3-p205.
**************************************************************************
checking for vm_core.h... yes
checking for version.h... yes
creating Makefile

current directory: C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/debase-0.2.2/ext
make "DESTDIR=" clean
      1 [main] make (12368) C:\Ruby24-x64\msys64\usr\bin\make.exe: *** fatal error - cygheap base mismatch detected - 0x41BF410/0x37BF410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
      0 [main] make 8532 fork: child -1 - forked process 12368 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
make: fork: Resource temporarily unavailable

current directory: C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/debase-0.2.2/ext
make "DESTDIR="
      1 [main] make (668) C:\Ruby24-x64\msys64\usr\bin\make.exe: *** fatal error - cygheap base mismatch detected - 0x376F410/0x41FF410.
This problem is probably due to using incompatible versions of the cygwin DLL.
Search for cygwin1.dll using the Windows Start->Find/Search facility
and delete all but the most recent version.  The most recent version *should*
reside in x:\cygwin\bin, where 'x' is the drive on which you have
installed the cygwin distribution.  Rebooting is also suggested if you
are unable to find another cygwin DLL.
      0 [main] make 11272 fork: child -1 - forked process 668 died unexpectedly, retry 0, exit code 0xC0000142, errno 11
make: fork: Resource temporarily unavailable

make failed, exit code 2

Gem files will remain installed in C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/debase-0.2.2 for inspection.
Results logged to C:/Ruby24-x64/lib/ruby/gems/2.4.0/extensions/x64-mingw32/2.4.0/debase-0.2.2/gem_make.out
C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:92:in `run'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:52:in `block in make'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:44:in `each'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:44:in `make'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/ext_conf_builder.rb:63:in `block in build'
  C:/Ruby24-x64/lib/ruby/2.4.0/tempfile.rb:295:in `open'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/ext_conf_builder.rb:31:in `build'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:164:in `block (2 levels) in build_extension'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:163:in `chdir'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:163:in `block in build_extension'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:162:in `synchronize'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:162:in `build_extension'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:201:in `block in build_extensions'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:198:in `each'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/ext/builder.rb:198:in `build_extensions'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/installer.rb:764:in `build_extensions'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/installer.rb:304:in `install'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/resolver/specification.rb:97:in `install'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/request_set.rb:166:in `block in install'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/request_set.rb:156:in `each'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/request_set.rb:156:in `install'
  C:/Ruby24-x64/lib/ruby/2.4.0/rubygems/dependency_installer.rb:405:in `install'
  mkrf_conf.rb:35:in `rescue in <main>'
  mkrf_conf.rb:28:in `<main>'

rake failed, exit code 1

Gem files will remain installed in C:/Ruby24-x64/lib/ruby/gems/2.4.0/gems/ruby-debug-ide-0.6.1 for inspection.
Results logged to C:/Ruby24-x64/lib/ruby/gems/2.4.0/extensions/x64-mingw32/2.4.0/ruby-debug-ide-0.6.1/gem_make.out
```

When searching for a cygwin1.dll - there is no result. 

## Steps to reproduce
Run the Installer and follow the instructions, until the msys2 installer is loaded

## What's the output from `ridk version`?
```
---
ruby:
  path: C:/Ruby24-x64
  version: 2.4.4
  platform: x64-mingw32
ruby_installer:
  package_version: 2.4.4-1
  git_commit: 21c8dc3
msys2:
  path: C:\Ruby24-x64\msys64
cc: x86_64-w64-mingw32-gcc (Rev1, Built by MSYS2 project) 7.3.0
sh: GNU bash, Version 4.4.19(2)-release (x86_64-pc-msys)
os: Microsoft Windows [Version 10.0.16299.248]
```

**Comment by larskanis** [comment]
> os: Microsoft Windows [Version 10.0.16299.248]

This version should work.

> When searching for a cygwin1.dll - there is no result.

The error message comes from Cygwin, but wasn't changed for MSYS2. The file is actually called `msys-2.0.dll` . You could follow the error message and search for more instances of this file, however I guess there's nothing wrong with your installations, but there is some kind of anti virus scanner or other realtime security protection, which causes MSYS2 to fail.

Please try RubyInstaller with AV temporary disabled.

**Comment by entwanderer** [comment]
Thank you for your feedback.

I will check your tip with the av scanner as soon as my administrators are in good mood.

**Comment by larskanis** [comment]
@entwanderer Did you have got any success with the AV scanner regarding the above issue?

**Comment by entwanderer** [comment]
@larskanis thank you for your query.

It indeed seems that the AV is the problem. Probably combined with the ASLR security feature that ships with Windows 10. But due organisational issues in our big company we decided for a work around that we could implement quickly. We installed it on a Windows 7 machine with different access rights / AV and copied the installed result to the Windows 10 machine.
